### PR TITLE
add a player dash

### PR DESCRIPTION
**Pull request by TheMadHatter01**

### CODE REVIEW

File: scenes/player/base_state.gd [hunk 1]
```diff
@@ -38,5 +38,13 @@ func physics_process(delta: float):
 		_state_machine.transition_to(_state_machine.State.JUMP)
 		_state_machine.get_current_state()._update(delta)
 		return
+	if (
+		Input.is_action_just_pressed("dash")
+		and _state_machine.current_state != _state_machine.State.DASH
+		and _player.can_dash == true
+	):
+		_state_machine.transition_to(_state_machine.State.DASH)
+		_state_machine.get_current_state()._update(delta)
```
Comment (AWarmCupOfTea) [inline scenes/player/base_state.gd:47]: I would suggest to make use of `_enter` function in Dash. Similar to how it is done in Jump.

File: scenes/player/dash.gd [hunk 2]
```diff
@@ -0,0 +1,47 @@
+class_name Dash
+extends BaseState
+
+onready var dash_timer = $DashTimer
+onready var dash_duration = _player.DASH_DURATION
+
+signal dash_ended
+
+
+func enter():
```
Comment (AWarmCupOfTea) [inline scenes/player/dash.gd:10]: The parent function starts with an underscore, same for exit.

File: scenes/player/dash.gd [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+
+func exit():
+	.exit()
+
+
+func _update(delta: float):
+	._update(delta)
+
+	if Input.is_action_just_pressed("dash") and !is_dashing() and _player.can_dash == true:
```
Comment (AWarmCupOfTea) [inline scenes/player/dash.gd:21]: I think if the player is already in the Dash state, then dash was already pressed anyway.

File: scenes/player/dash.gd [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		start_dash(dash_duration)
+		var dash_direction = _player.get_move_input_direction()
+		_player.velocity.x = _player.DASH_SPEED * dash_direction
+	_player.move_player(delta)
+
+
+func start_dash(duration):
```
Comment (AWarmCupOfTea) [inline scenes/player/dash.gd:28]: Please add `: float` type hint.

File: scenes/player/player.gd [hunk 5]
```diff
@@ -23,12 +23,18 @@ const COYOTE_TIME_SECS := 0.1
...
+const DASH_DURATION = 0.3
 
 # Consts end
 
 var velocity := Vector2.ZERO
+var dash_delay = 1
+var can_dash = true
 
 onready var _state_machine = $StateMachine
+onready var dash = $StateMachine/BaseState/Dash
```
Comment (AWarmCupOfTea) [inline scenes/player/player.gd:37]: Is this needed? I don't think it is used.

**Review by AWarmCupOfTea — CHANGES_REQUESTED** [review]

File: scenes/player/dash.gd (hunk 3 above)
Comment (TheMadHatter01) [reply scenes/player/dash.gd:21]: yah, I didn't realize that I could use the _enter func, for some reason i thought that was for something else😂

**Review by AWarmCupOfTea — APPROVED** [review]

**Review by AWarmCupOfTea — APPROVED** [review]

**Pull request merged**